src/jit/morph.cpp

```diff
@@ -6940,263 +6945,236 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
-    calleeStackSize        = calleeStackSlots * TARGET_POINTER_SIZE;
-    size_t callerStackSize = info.compArgStackSize;
+    // If we reached here means that callee has only those argument types which can be passed in
+    // a register and if passed on stack will occupy exactly one stack slot in out-going arg area.
+    // If we are passing args on stack for the callee and it has more args passed on stack than
+    // the caller, then fast tail call cannot be performed.
+    //
+    // Note that the GC'ness of on stack args need not match since the arg setup area is marked
+    // as non-interruptible for fast tail calls.
 
-    bool hasStackArgs = false;
+    bool calleeHasStackArgs    = false;
 
+    // LowerFastTail has logic to swap incoming and outgoing args that are on the stack.
+    // LowerFastTailCall currently only support swapping/copying 1 slot
+    // stack argument. All arguments are assumed to be 1 slot size. If there are no calleeStackArgs
+    // then we will not take that code path. That being said, if the caller has stack args
+    // but the callee does not we will be able to fast tail call.
     if (callerStackSize > 0 || calleeStackSize > 0)
     {
-        hasStackArgs = true;
+        calleeHasStackArgs = true;
     }
 
-    // Go the slow route, if it has multi-byte params. This is an implementation
-    // limitatio see https://github.com/dotnet/coreclr/issues/12644.
-    if (hasMultiByteStackArgs)
-    {
-        reportFastTailCallDecision("Will not fastTailCall hasMultiByteStackArgs", callerStackSize, calleeStackSize);
-        return false;
-    }
+#ifdef WINDOWS_AMD64_ABI
+    callerIntArgRegCount = callerIntArgRegCount + calleeFloatArgRegCount;
+
 
     // x64 Windows: If we have more callee registers used than MAX_REG_ARG, then
     // make sure the callee's incoming arguments is less than the caller's
-    if (hasStackArgs && (nCalleeArgs > nCallerArgs))
+    if (calleeHasStackArgs && (nCalleeArgs > nCallerArgs))
     {
-        reportFastTailCallDecision("Will not fastTailCall hasStackArgs && (nCalleeArgs > nCallerArgs)", callerStackSize,
+        reportFastTailCallDecision("Will not fastTailCall calleeHasStackArgs && (nCalleeArgs > nCallerArgs)", callerStackSize,
                                    calleeStackSize);
         return false;
     }
 
 #elif (defined(_TARGET_AMD64_) && defined(UNIX_AMD64_ABI)) || defined(_TARGET_ARM64_)
 
+    // Because there is an assumption in LowerFastTailCall that all args are 
+    // 1 slot size, we require the incoming caller args to also not contain a 
+    // largerThanOneStackSlotSizedStruct. Therefore, loop over the caller's args
+    // and check for any by value structs > one stack slot.
+
+#if FEATURE_MULTIREG_ARGS
+    int varNum = 0;
+    for (LclVarDsc* varDsc = this->lvaTable; varNum < this->lvaCount; ++varNum, ++varDsc)
+    {
+        // Is this variable a parameter?
+
+        if (!varDsc->lvIsParam)
+        {
+            continue;
```

Comment:
I believe it's (still) the case that all parameters are located at the beginning of the `lvaTable`, so that this could be `break;`. In looking at this, I find that there are places that assume this, e.g. here: https://github.com/dotnet/coreclr/blob/master/src/jit/codegencommon.cpp#L1942 (where it breaks out of the loop when it finds a non-reg-arg) and here: https://github.com/dotnet/coreclr/blob/master/src/jit/codegencommon.cpp#L6791 (where it only iterates to `compArgsCount`).
But here: https://github.com/dotnet/coreclr/blob/master/src/jit/codegencommon.cpp#L3449 and here: https://github.com/dotnet/coreclr/blob/master/src/jit/codegencommon.cpp#L4585 we continue to loop.
So usage is not consistent, but I'm pretty sure it's a valid invariant.